File: backend/src/main/java/com/beej/product/service/ProductService.java
package com.beej.product.service;

import com.beej.product.dto.ProductResponseDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.math.BigDecimal;
import java.util.List;

public interface ProductService {
    
    Page<ProductResponseDTO> getProducts(Long categoryId, String search, BigDecimal minPrice, BigDecimal maxPrice, Pageable pageable);
    
    ProductResponseDTO getProductById(Long id);
    
    List<ProductResponseDTO> getFeaturedProducts();
    
    Page<ProductResponseDTO> searchProducts(String query, Pageable pageable);
    
    List<ProductResponseDTO> getNewArrivals();
    
    List<ProductResponseDTO> getTrendingProducts();
}

File: frontend/src/services/productService.js
import api from '../api/axios'

const productService = {
  // Get all products with filters and pagination
  getProducts: async (params = {}) => {
    const response = await api.get('/products', { params })
    return response.data.data
  },

  // Get product by ID
  getProductById: async (id) => {
    const response = await api.get(`/products/${id}`)
    return response.data.data
  },

  // Get all categories
  getCategories: async () => {
    const response = await api.get('/categories')
    return response.data.data
  },

  // Get products by category
  getProductsByCategory: async (categoryId, params = {}) => {
    const response = await api.get(`/categories/${categoryId}/products`, { params })
    return response.data.data
  },

  // Search products
  searchProducts: async (query, params = {}) => {
    const response = await api.get('/products/search', { 
      params: { q: query, ...params } 
    })
    return response.data.data
  },

  // Get featured products
  getFeaturedProducts: async () => {
    const response = await api.get('/products/featured')
    return response.data.data
  },

  // Get trending products
  getTrendingProducts: async () => {
    const response = await api.get('/products/trending')
    return response.data.data
  },

  // Get new arrivals
  getNewArrivals: async () => {
    const response = await api.get('/products/new-arrivals')
    return response.data.data
  },

  // Get related products
  getRelatedProducts: async (productId) => {
    const response = await api.get(`/products/${productId}/related`)
    return response.data.data
  },

  // Get product reviews
  getProductReviews: async (productId, params = {}) => {
    const response = await api.get(`/products/${productId}/reviews`, { params })
    return response.data.data
  },

  // Add product review
  addProductReview: async (productId, reviewData) => {
    const response = await api.post(`/products/${productId}/reviews`, reviewData)
    return response.data.data
  },
}

export default productService
